spatialdata/_core/_spatialdata.py

```diff
@@ -591,62 +695,177 @@ def add_shapes(
+        store = parse_url(file_path, mode="w").store
+
+        root = zarr.group(store=store, overwrite=overwrite)
+        store.close()
+
+        # old code to support overwriting the backing file
+        # if target_path is None:
+        #     self.path = str(file_path)
+        # else:
+        #     self.path = str(tmp_zarr_file)
+        self.path = str(file_path)
         try:
             if len(self.images):
                 root.create_group(name="images")
-                # add_image will delete and replace the same key in self.images, so we need to make a copy of the
+                # add_image_in_memory will delete and replace the same key in self.images, so we need to make a copy of the
                 # keys. Same for the other elements
                 keys = list(self.images.keys())
-                for el in keys:
-                    self.add_image(
-                        name=el, image=self.images[el], storage_options=storage_options, _add_in_memory=False
+                from spatialdata._io.read import _read_multiscale
+
+                for name in keys:
+                    elem_group = self._init_add_element(name=name, element_type="images", overwrite=overwrite)
+                    write_image(
+                        image=self.images[name],
+                        group=elem_group,
+                        name=name,
+                        storage_options=storage_options,
                     )
 
+                    # reload the image from the Zarr storage so that now the element is lazy loaded, and most importantly,
+                    # from the correct storage
+                    element_path = Path(self.path) / "images" / name
+                    image = _read_multiscale(str(element_path), raster_type="image")
+                    self._add_image_in_memory(name=name, image=image, overwrite=True)
+
             if len(self.labels):
                 root.create_group(name="labels")
                 keys = list(self.labels.keys())
-                for el in keys:
-                    self.add_labels(
-                        name=el, labels=self.labels[el], storage_options=storage_options, _add_in_memory=False
+                from spatialdata._io.read import _read_multiscale
+
+                for name in keys:
+                    elem_group = self._init_add_element(name=name, element_type="labels", overwrite=overwrite)
+                    write_labels(
+                        labels=self.labels[name],
+                        group=elem_group,
+                        name=name,
+                        storage_options=storage_options,
                     )
 
+                    # reload the labels from the Zarr storage so that now the element is lazy loaded, and most importantly,
+                    # from the correct storage
+                    element_path = Path(self.path) / "labels" / name
+                    labels = _read_multiscale(str(element_path), raster_type="labels")
```

Comment:
same here, I'd ttry to be consistent everywhere

Reply:
As above, I removed this cast to `str`.